File: ClientChatter/ConnectionManager.cs
using System;
using System.Threading.Tasks;
using ClientChatter.Model;
using Microsoft.AspNetCore.SignalR.Client;

namespace ClientChatter
{
  public class ConnectionManager : IConnectionManager
  {
    private readonly HubConnection _connection;

    public ConnectionManager(HubConnection connection)
    {
      _connection = connection;
    }

    public void RegisterEventHandlers()
    {
      Console.WriteLine($"Registering Reconnecting Event");
      //When a connection is lost , the HubConnection will transition to the HubConnectionState.Reconnecting state and fire
      //the Reconnecting event. This provides an opportunity to warn users that the connection has been lost 
      _connection.Reconnecting += Connection_Reconnecting;

      Console.WriteLine($"Registering Reconnected Event");
      //If the client successfully reconnects within its first four attempts, the HubConnection will transition back to the Connected state
      //and fire the Reconnected event. Since the connection looks entirely new to the server, a new ConnectionId will be provided to the
      //Reconnected event handlers.
      _connection.Reconnected += Connection_Reconnected;

      Console.WriteLine($"Registering Closed Event");
      //If the client doesn't successfully reconnect within its first four attempts, the HubConnection will transition to the Disconnected
      //state and fire the Closed event. This provides an opportunity to attempt to restart the connection 
      _connection.Closed += Connection_Closed;

    }

    //------------------------------------------------------------------------------------------------------------------------------------------------
    public void RegisterClientMethods()
    {
      //Define methods the server/hub calls. Must be specified but before starting the connection.

      Console.WriteLine($"Registering 'Welcome' Method Handler ");
      _connection.On<ChatRoomMessage>("Welcome", (message) =>
      {
        Console.WriteLine($"Message Received from hub");
        Console.WriteLine($"From: {message.UserName} - Sent: {message.TimeSent} - Message: {message.Message}");
      });


      Console.WriteLine($"Registering 'ReceiveMessage' Method Handler ");
      _connection.On<ChatRoomMessage>("ReceiveMessage", (message) =>
      {
        Console.WriteLine($"Message Received from hub");
        Console.WriteLine($"From: {message.UserName} - Sent: {message.TimeSent} - Message: {message.Message}");
      });
    }
    //------------------------------------------------------------------------------------------------------------------------------------------------

    //------------------------------------------------------------------------------------------------------------------------------------------------
    public async Task StartConnection()
    {
      // Starts a connection to the server.
      Console.WriteLine($"Starting Hub Connection ");
      try
      {
        await _connection.StartAsync();
      }
      catch (Exception ex)
      {
        Console.WriteLine($"Error when starting connection - {ex}");
      }
    }
    //------------------------------------------------------------------------------------------------------------------------------------------------

    //------------------------------------------------------------------------------------------------------------------------------------------------
    //The client has SendAsync which is fire - and - forget and InvokeAsync which waits for a completion message from the server.https://github.com/aspnet/SignalR/issues/1300
    //Invokes a hub method on the server using the specified method name.
    public async Task Send_MessageToAll_ToServer()
    {
      Console.WriteLine($"Call to 'SendMessageToAll' method on hub/server ");
      try
      {
        await _connection.InvokeAsync("SendMessageToAll", "Hi");
      }
      catch (Exception ex)
      {
        Console.WriteLine($"Exception on Send Message: {ex}");
      }
    }
    //------------------------------------------------------------------------------------------------------------------------------------------------

    //------------------------------------------------------------------------------------------------------------------------------------------------
    private static Task Connection_Reconnecting(Exception arg)
    {
      Console.WriteLine($"Reconnecting");
      return Task.CompletedTask;
    }
    //------------------------------------------------------------------------------------------------------------------------------------------------

    //------------------------------------------------------------------------------------------------------------------------------------------------
    private static Task Connection_Reconnected(string arg)
    {
      Console.WriteLine($"Reconnected with connection id {arg} ");
      return Task.CompletedTask;
    }
    //------------------------------------------------------------------------------------------------------------------------------------------------

    //------------------------------------------------------------------------------------------------------------------------------------------------
    private static Task Connection_Closed(Exception arg)
    {
      Console.WriteLine($"Connection Closed ");
      //await Task.Delay(new Random().Next(0, 5) * 1000);
      //await _connection.StartAsync();
      return Task.CompletedTask;
    }
    //------------------------------------------------------------------------------------------------------------------------------------------------
  }
}

File: ChattingServer/wwwroot/js/ConnectionManager.js
/// <reference path="../lib/signalr/signalr.js" />
/// <reference path="../js/Models/ChatRoomMessage.js"/>

"use strict";

function ConnectionManager() {

    //---------------------------------------------------------------------------------------------------------------------------------------
    //Private Methods and Members

    // Create Hub Connection
    var _clientChatterConnection = new signalR.HubConnectionBuilder()
        .withUrl("https://localhost:44333/chatroom")
        .build();


    //Private Functions

    //Move this DisplayMessage Code Out!!!!
    var displayMessage = function (message) {
        var eleTextAreaChatHistory = document.getElementById('chatMsgTextAreaId');
        eleTextAreaChatHistory.value += JSON.stringify(message);
    }


    var handleWelcome = function (message) {
        console.log(`Welcome Received from hub`);
        console.log(`From: ${message.UserName} - Sent: ${message.TimeSent} - Message: ${message.Message}`);
        displayMessage(message);
    }

    var handleReceiveMessage = function (message) {
        console.log(`Received Message Received from hub`);
        console.log(`From: ${message.UserName} - Sent: ${message.TimeSent} - Message: ${message.Message}`);
        displayMessage(message);
    }

    var handleOnClosedEvent = function () {
        console.log(`Received On Closed Event`);
    }



    //---------------------------------------------------------------------------------------------------------------------------------------


    //---------------------------------------------------------------------------------------------------------------------------------------
    //Public Methods
    //Start the Hub Connection
    this.StartConnection = function () {
        _clientChatterConnection.start()
            .catch((err) => console.error(err.toString()));
    }

    this.CloseConnection = function() {
        _clientChatterConnection.stop()
            .catch((err) => console.error(err.toString()));
    }


    this.RegisterClientMethods = function () {
        //Define methods the server/hub calls. Must be specified but before starting the connection.

        console.log(`Registering 'Welcome' Method Handler `);
        _clientChatterConnection.on('Welcome', handleWelcome);

        console.log(`Registering 'ReceiveMessage' Method Handler `);
        _clientChatterConnection.on('ReceiveMessage', handleReceiveMessage);
    }

    this.RegisterEventHandlers = function () {

        console.log(`Registering Closed Event`);
        _clientChatterConnection.onclose(handleOnClosedEvent);

    }

    this.Send_MessageToAll_ToServer = function (fromUsername, message) {
        console.log(`Call to "SendMessageToAll" method on hub/server`);

        var currentTime = new Date();
        var chatRoomMessage = new ChatRoomMessage(fromUsername, message, currentTime);

        _clientChatterConnection.invoke('SendMessageToAll', chatRoomMessage)
            .catch((err) => console.log(err.toString()));

    }

    this.Send_MessageToGroup_ToServer = function (groupName, fromUsername, message) {
        console.log(`Call to "SendMessageToUserGroup" method on hub/server`);

        var currentTime = new Date();
        var chatRoomMessage = new ChatRoomMessage(fromUsername, message, currentTime);

        _clientChatterConnection.invoke('SendMessageToUserGroup', groupName, chatRoomMessage)
            .catch((err) => console.log(err.toString()));

    }
    //---------------------------------------------------------------------------------------------------------------------------------------
}
